### PR TITLE
Dockerfile for Yosys

### PR DESCRIPTION
**Pull request by abdelrahmanhosny**
This pull request proposes to dockerize Yosys suite. To build the Docker image: `docker build -t yosys .`. To use the image, mount the directory where you have your RTL and library files to `/data` inside the image. There are two options to run yosys:

- Use the interactive shell: `docker run -it -v <data-directory>:/data yosys`. Read files from `/data` directory. For example: `read_verilog /data/design.v`
- Use the yosys commands from a file: `docker run -v <data-directory>:/data yosys /data/synth.ys`, where `synth.ys` should be located in the host data directory as well.

To try without building the image locally, use image name `openroad/yosys` on Docker Hub: https://hub.docker.com/r/openroad/yosys

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,47 @@
+FROM ubuntu:16.04 as builder
```
Comment (q3k) [inline Dockerfile]: Can we bump this to 18.04? That's the latest LTS.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    clang \
+    bison \
+    flex \
+    libreadline-dev \
+    gawk \
+    tcl-dev \
+    libffi-dev \
+    git \
+    graphviz \
+    xdot \
```
Comment (q3k) [inline Dockerfile]: Do we need graphviz/xdot when building?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    pkg-config \
+    python3
+
+COPY . /
+RUN make && \
+    make install
+
+
+FROM ubuntu:16.04
+RUN apt-get update && apt-get install -y clang \
```
Comment (q3k) [inline Dockerfile]: Do we need clang, bison, flex, libreadline-dev, tcl-dev, libffi-dev, git, pkg-config at runtime?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    flex \
+    libreadline-dev \
+    gawk \
+    tcl-dev \
+    libffi-dev \
+    git \
+    graphviz \
+    xdot \
+    pkg-config \
+    python3
```
Comment (q3k) [inline Dockerfile]:
```
    python && \
    rm -rf /var/lib/apt/lists
```

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+COPY --from=builder /yosys /build/yosys
+COPY --from=builder /yosys-abc /build/yosys-abc
+COPY --from=builder /yosys-config /build/yosys-config
+COPY --from=builder /yosys-filterlib /build/yosys-filterlib
+COPY --from=builder /yosys-smtbmc /build/yosys-smtbmc
+
+ENV PATH /build:$PATH
+
+RUN mkdir /data
```
Comment (q3k) [inline Dockerfile]: This should be up to the end-user of the Dockerfile, no?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+WORKDIR /data
+
+ENTRYPOINT ["yosys"]
```
Comment (q3k) [inline Dockerfile]:
```
RUN useradd -m yosys
USER yosys
```

File: Dockerfile (hunk 2 above)
Comment (q3k) [reply Dockerfile]: (this likely pulls in X11, which we don't want)

File: Dockerfile (hunk 5 above)
Comment (q3k) [reply Dockerfile]: I mean, even for `-v foo:/data`, you don't need to create this directory first in the Dockerfile.

**Comment by q3k** [comment]
Thanks! Let's sort out the above nitpicks, and I'd love to merge it. I'll also set up a yosys orga on dockerhub.

File: Dockerfile (hunk 1 above)
Comment (q3k) [reply Dockerfile]: (also the runtime container below)

Comment (abdelrahmanhosny) [reply Dockerfile]: fixed

File: Dockerfile (hunk 2 above)
Comment (abdelrahmanhosny) [reply Dockerfile]: removed.

File: Dockerfile (hunk 3 above)
Comment (abdelrahmanhosny) [reply Dockerfile]: I only kept `libreadline-dev` and `tcl-dev` and removed the rest. They give the following errors when removed:
`error while loading shared libraries`

File: Dockerfile (hunk 4 above)
Comment (abdelrahmanhosny) [reply Dockerfile]: fixed.

File: Dockerfile (hunk 5 above)
Comment (abdelrahmanhosny) [reply Dockerfile]: You are right. I removed it.

File: Dockerfile (hunk 6 above)
Comment (abdelrahmanhosny) [reply Dockerfile]: added.

**Comment by abdelrahmanhosny** [comment]
Thank you Serge for pointing out these improvements. I addressed them accordingly. The image size is now `553MB`. Let me know if you see more room for improvement.
I tested the resulting image on a sample design and it worked as expected.

**Comment by XVilka** [comment]
Have you considered to use Alpine linux as container base instead? It will help to reduce image size substantially. See for example https://github.com/gliderlabs/docker-alpine

**Comment by abdelrahmanhosny** [comment]
I tried `alpine` and the resulting image is `420MB` in size. I can push the Dockerfile if you think it is better and would prefer it over the `ubuntu` image.

**Comment by q3k** [comment]
Let's get this merged as is now. We can then consider moving to an Alpine-based container.

**Pull request merged**